feat: add support for the assertions endpoint

ashuntu opened the pull request:
Snaps currently can't be obtained by their ID, but they can be via the [`assertions` snapd endpoint](https://snapcraft.io/docs/snapd-rest-api#heading--remote-assertion). The endpoint isn't currently implemented in this client, so this adds support for it, ~~as well as a convenience function to find a snap by its ID~~.

Finding a snap by ID is going to be required by [App Center](https://github.com/ubuntu/app-center)'s gaming tab, since I'll need to be able to obtain a snap from its [rating](https://github.com/ubuntu/app-center-ratings) (which only contains the ID), similar to what's already described in https://github.com/canonical/snapd.dart/issues/99.

Fixes https://github.com/canonical/snapd.dart/issues/99.

robert-ancell commented on lib/src/snapd_client.dart:
```diff
@@ -1418,6 +1475,45 @@ class SnapdClient {
     return snaps;
   }
 
+  /// Convienience function to find a snap by its ID. Under the hood, this
```
Convienience > Convenience

robert-ancell commented on lib/src/snapd_client.dart:
```diff
@@ -1565,6 +1661,17 @@ class SnapdClient {
   }
 
+  /// Does a syncronous request to snapd without parsing the response.
+  Future<String> _getSyncRaw(String path,
```
Refactor _getSync and _getSyncRaw so they share a common function.

robert-ancell commented on test/snapd_test.dart:
```diff
@@ -522,6 +522,7 @@ class MockSnapdServer {
 
     var method = request.method;
     var path = request.uri.path;
+
```
This whitespace seems accidentally added.

robert-ancell commented on test/snapd_test.dart:
```diff
@@ -793,6 +796,43 @@ class MockSnapdServer {
     _writeSyncResponse(request.response, snaps);
   }
 
+  void _processFindById(HttpRequest request) async {
```
The tests try not to have hard-coded behaviour like this - instead make a MockAssertionObject that is provided in the MockSnapdConstructor (in a map or similar). Then the contents of the assertion will be in the test.

robert-ancell commented:
I changed this to a draft as you said the endpoint isn't implemented - do you mean snapd doesn't support these calls or are you referring to something else?

ashuntu commented:
> I changed this to a draft as you said the endpoint isn't implemented - do you mean snapd doesn't support these calls or are you referring to something else?

Ah yeah, perhaps my original description was misleading. I just meant that this snapd client didn't yet implement the assertions endpoint (which is available and working in snapd).

robert-ancell commented:
It's probably worth splitting this into two PRs - adding the support for the endpoint should be simple, the helper function probably needs more thought / review.

ashuntu commented:
I removed the utility function, so this PR is just the assertion endpoint. I'll open another for the utility function.

I *should* have addressed your comments, but let me know if I missed anything.

robert-ancell commented on lib/src/snapd_client.dart:
```diff
@@ -1418,6 +1475,21 @@ class SnapdClient {
 
+  /// List all assertions of the given [assertion] type.
+  /// If no [assertion] type is provided, lists available assertion types.
+  Future<Map<String, dynamic>> assertions(
```
This should be named getAssertions as per the naming of the other methods.

robert-ancell commented on lib/src/snapd_client.dart:
```diff
@@ -1418,6 +1475,21 @@ class SnapdClient {
+  /// List all assertions of the given [assertion] type.
+  /// If no [assertion] type is provided, lists available assertion types.
+  Future<Map<String, dynamic>> assertions(
+      {String? assertion, Map<String, String>? params}) async {
```
Instead of a generic params this should be a list of known params, e.g. `{String? series, bool remote = false, String ?snapId}` and then this method makes the map from these.

robert-ancell replied:
Ah, I wasn't aware of that. You should document what [params] is and either some of the possible params or link to some documentation about them.

robert-ancell replied:
And consider if it makes sense to add a helper method which does have the params set, e.g. `getAccountAssertions()`

ashuntu replied:
There are [quite a few assertion types](https://snapcraft.io/docs/snapd-rest-api#heading--assertions), but that could be theoretically possible.

robert-ancell replied:
Up to you if you think it would be worthwhile - just trying to reduce the changes of an app using this library from making a mistake and having to read a lot of documentation.

robert-ancell reviewed (changes requested):
Thanks for the update - just some API adjustments and it should be good to land.

robert-ancell reviewed (approved):
Looks good, thanks!
